nemo_skills/evaluation/top_answer_processor.py

```diff
@@ -0,0 +1,314 @@
+    # Convert all to Path objects
+    file_path, input_dir, output_dir = Path(file_path), Path(input_dir), Path(output_dir)
+
+    # Get the relative path from input_dir to the file
+    relative_path = file_path.relative_to(input_dir)
+
+    # Combine output_dir with the relative path
+    output_path = output_dir / relative_path
+
+    # Create parent directories if they don't exist
+    output_path.parent.mkdir(parents=True, exist_ok=True)
+
+    return output_path
+
+
+class ProcessMode(Enum):
+    FILL = "fill"
+    EXTRACT = "extract"
+
+
+class TopAnswerProcessor:
+    def __init__(self, cfg: ProcessTopAnswerConfig):
+        self.cfg = cfg
+        self._validate_cfg()
+
+    def _validate_cfg(self):
+        """Validate the config"""
+        cfg = self.cfg
+        # Check if the mode is valid
+        try:
+            self.process_mode = ProcessMode(cfg.mode)
+        except ValueError:
+            raise ValueError(f"Invalid mode: {cfg.mode}")
+        
+        # For fill mode, output_dir is required
+        if self.process_mode == ProcessMode.FILL:
+            if cfg.output_dir is None:
+                raise ValueError("output_dir is required when mode is fill")
+        
+        # Check whether the input_dir and output_dir are valid
+        if not Path(cfg.input_dir).exists():
+            raise ValueError(f"Input directory does not exist: {cfg.input_dir}")
+        if cfg.output_dir is not None and not Path(cfg.output_dir).exists():
+            raise ValueError(f"Output directory does not exist: {cfg.output_dir}")
+
+            
+    def __enter__(self):
+        """Setup input and output file handles"""
+        cfg = self.cfg
+        # Process the input files
+        self.input_files = list(unroll_files(cfg.input_files, parent_dir=cfg.input_dir))
+        self.input_file_handles = [open(file, "rt", encoding="utf-8") for file in self.input_files]
+        
+        if self.cfg.require_num_files is not None:
+            if len(self.input_file_handles) != cfg.require_num_files:
+                raise ValueError(f"Expected {cfg.require_num_files} files, found {len(self.input_file_handles)}")
+
+        if self.process_mode == ProcessMode.FILL:
+            # Create output files and their handles with the same relative paths as the input files
+            output_files = [
+                map_to_output_path(file, cfg.input_dir, cfg.output_dir)
+                for file in unroll_files(cfg.input_files, parent_dir=cfg.input_dir)
+            ]
+            self.output_file_handles = [open(file, "wt", encoding="utf-8") for file in output_files]
+        
+        elif self.process_mode == ProcessMode.EXTRACT:
+            if cfg.output_dir is None:
+                cfg.output_dir = cfg.input_dir
+            
+            # Determine the file suffix based on the use of RM scores
+            if cfg.use_majority_rm_score:
+                file_suffix = "-majority-rm.jsonl"
```

Comment:
I think this might make it harder to reuse this script as part of a pipeline since you'd need to be changing the name there as well. I would instead include another "fill_mode" entry in the inner dictionary on each line, so that we can refer to that later on. Will also be easier to track as it will be carried over to any subsequent generations / processing while file name will be much harder to find later on

Reply:
Good point. Let me introduce the `fill_mode` entry. 
For the extract mode, let's just refer the file as `output-agg.jsonl`?

Reply:
sure